Switch to cmake part1

peterNordin opened the pull request:
This is part one of switch over to CMake, has been in progress for more than three years.
After adding DCP additional work is needed to be able to build release, but i will make a part2

This PR enables me to build all of Hopsan on Linux GCC, Windows MinGW and Windows MSVC 2019.
Including with DCP. But I do not know if it actually works in all these cases. Hopsan at least starts.

Deb build has been migrated to CMake. But flatpak/snap and Windows builds remain.

peterNordin committed:
Switch to CMake in deb packaging

peterNordin committed:
Switch to CMake in deb packaging

peterNordin committed:
Reuse existing clone of code to avoid redownload

peterNordin committed:
Reuse existing clone of code to avoid redownload

peterNordin committed:
Add missing doxygen/graphviz to deb control files

peterNordin committed:
Add missing doxygen/graphviz to deb control files

peterNordin committed:
Optionally use cache directory for dependency downloads

To prevent re-downloading after cleaning code during test builds

peterNordin committed:
Optionally use cache directory for dependency downloads

To prevent re-downloading after cleaning code during test builds

peterNordin committed:
Use cache dir for downloads, and keep temporary stage dir

Keep the dir to avoid fetching all code every time

peterNordin committed:
Use cache dir for downloads, and keep temporary stage dir

Keep the dir to avoid fetching all code every time

peterNordin committed:
Adjust cmake policys for older CMake support

peterNordin committed:
Adjust cmake policys for older CMake support

peterNordin committed:
Fix msgpackc workaround

peterNordin committed:
Fix msgpackc workaround

peterNordin committed:
Fix compatibility with older cmake

peterNordin committed:
Fix compatibility with older cmake

peterNordin committed:
Add missing ARCHIVE destinations for static libs

peterNordin committed:
Add missing ARCHIVE destinations for static libs

peterNordin committed:
Prefer persistent cache dir in workspace

peterNordin committed:
Prefer persistent cache dir in workspace

peterNordin committed:
Remove old minimum cmake version, let main file decide

Removes warning spam on newer CMake versions

peterNordin committed:
Remove old minimum cmake version, let main file decide

Removes warning spam on newer CMake versions

peterNordin committed:
Remove experimental Warning

peterNordin committed:
Remove experimental Warning

peterNordin committed:
Update fmi4c and use the new cmake configuration

peterNordin committed:
Update fmi4c and use the new cmake configuration

peterNordin committed:
Cleanup Linux build scripts

peterNordin committed:
Cleanup Linux build scripts

peterNordin committed:
Update setupDCPlib and use cmake prefix path

Fix dcplib build on Linux and with MSVC

peterNordin committed:
Update setupDCPlib and use cmake prefix path

Fix dcplib build on Linux and with MSVC

peterNordin committed:
Simplify CMake DCP lookup

Use alias target compatible with older cmake

peterNordin committed:
Simplify CMake DCP lookup

Use alias target compatible with older cmake

peterNordin committed:
Install dcp dependencies zlib, libzip and xerces

peterNordin committed:
Install dcp dependencies zlib, libzip and xerces

peterNordin committed:
Update gitignore

peterNordin committed:
Update gitignore

peterNordin committed:
Cleanup hopsanremote CMake files

peterNordin committed:
Cleanup hopsanremote CMake files

peterNordin committed:
Switch appveyor MinGW build to CMake

peterNordin committed:
Switch appveyor MinGW build to CMake

peterNordin committed:
Update to 2019 appveyor image

Hopefully it has tar installed in the system

peterNordin committed:
Update to 2019 appveyor image

Hopefully it has tar installed in the system

robbr48 commented on hopsandcp/CMakeLists.txt:
```diff
@@ -1,19 +1,19 @@
 endif()
 
-set(CMAKE_AUTOMOC ON)
+include(${CMAKE_CURRENT_LIST_DIR}/../dependencies/dcplib.cmake)
```
We should not include this, since DCPLib is header-only and does no longer require linking.

robbr48 commented on hopsandcp/CMakeLists.txt:
```diff
@@ -1,19 +1,19 @@
+#include(${CMAKE_CURRENT_LIST_DIR}/../dependencies/xerces.cmake)
 
-# include(${CMAKE_CURRENT_LIST_DIR}/../dependencies/xerces.cmake)
 # include(${CMAKE_CURRENT_LIST_DIR}/../dependencies/libzip.cmake)
```
We need to include libzip.cmake, however, because we must link to libzip directly below.

peterNordin replied:
I moved the libzip setup into dcplib.cmake since it was only a dependency to dcplib, ..... I think.

robbr48 commented on hopsandcp/CMakeLists.txt:
```diff
@@ -25,22 +25,23 @@ add_library(hopsandcp STATIC ${srcfiles} ${headerfiles})
-  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../dependencies/libzip/include>
-  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../dependencies/asio-code/include>
-  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../dependencies/xerces/include>
+  # $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../dependencies/dcplib-code/include/bluetooth>
```
These include directories are needed, because we need to link to the headers in DCPLib, libzip, Asio and Xerces. But perhaps they should can be in dcplib.cmake if we remove all stuff with linking to the library.

robbr48 commented on hopsandcp/CMakeLists.txt:
```diff
@@ -25,22 +25,23 @@ add_library(hopsandcp STATIC ${srcfiles} ${headerfiles})
-    target_link_libraries(hopsandcp hopsancore libzip -lws2_32)
-else()
-    target_link_libraries(hopsandcp hopsancore libzip)
+target_link_libraries(hopsandcp hopsancore DCPLib::Master DCPLib::Slave DCPLib::Zip DCPLib::Ethernet)
```
Remove: DCPLib::Master, DCPLib::Slave and DCPLib::Zip
Add: libzip

peterNordin replied:
Those are the targets defined by the dcplib project, so I wanted to "link" (depend) on them instead of explicitly pointing to the header directories. There may be more properties in these targets (compiler definitions and such things) that may be required. I think that DCPLib::Zip depends on libzip

robbr48 reviewed (changes requested):
Seems to work except for the DCPLib. It no is no longer compiled and linked since it is header-only. See separate comments in hopsandcp/CMakeLists.txt.

The conflict is just because I fixed the same typo, so it is very easy to fix.

robbr48 committed:
Upgrade build server to Windows 2025

robbr48 committed:
Upgrade build server to Windows 2025

peterNordin committed:
Update qwt build patch

peterNordin committed:
Update qwt build patch

peterNordin committed:
Add zlib and lib zip to HDF5 build

Must be present for MSVC build

peterNordin committed:
Add zlib and lib zip to HDF5 build

Must be present for MSVC build

peterNordin committed:
Update xerces build and import on Windows

peterNordin committed:
Update xerces build and import on Windows

peterNordin committed:
Update discount build for msvc

peterNordin committed:
Update discount build for msvc

peterNordin committed:
Use /bigobj for MSVC in default library

peterNordin committed:
Use /bigobj for MSVC in default library

peterNordin committed:
Update fmi4c with new zlib with new cmake support

peterNordin committed:
Update fmi4c with new zlib with new cmake support

peterNordin committed:
Cleanup, GUI does not depend on Xerces and zlib directly

peterNordin committed:
Cleanup, GUI does not depend on Xerces and zlib directly

peterNordin committed:
Downgrade appveyor settings to 2017

- Qt 5.15 was removed so downgrading
- Adjust build order

peterNordin committed:
Downgrade appveyor settings to 2017

- Qt 5.15 was removed so downgrading
- Adjust build order

peterNordin committed:
Set DCMAKE_POLICY_VERSION_MINIMUM=3.5 for dependencies

peterNordin committed:
Set DCMAKE_POLICY_VERSION_MINIMUM=3.5 for dependencies

peterNordin committed:
Adjust buildorder

peterNordin committed:
Adjust buildorder

peterNordin committed:
Older cmake can not handle empty string argument

peterNordin committed:
Older cmake can not handle empty string argument

peterNordin committed:
Remove --fresh for older cmake

peterNordin committed:
Remove --fresh for older cmake

peterNordin committed:
Remove obsolete buster config

peterNordin committed:
Remove obsolete buster config

peterNordin committed:
Require python3

peterNordin committed:
Require python3

peterNordin committed:
Re-enable build with older qwt

peterNordin committed:
Re-enable build with older qwt

peterNordin committed:
replace plucky with questing

peterNordin committed:
replace plucky with questing

peterNordin committed:
Add missing header for new gcc

peterNordin committed:
Add missing header for new gcc

peterNordin committed:
Adjust build order and add local discount build

peterNordin committed:
Adjust build order and add local discount build

peterNordin committed:
Define USEWEBENGINE

peterNordin committed:
Define USEWEBENGINE

peterNordin committed:
Update windowsRelease script to use cmake and python 3

- Modernize to Pyton 3 (with AI help)
- Build Mingw using CMake
- MSVC build disabled, but can be added as an option later
- Various related changes and fixes

peterNordin committed:
Update windowsRelease script to use cmake and python 3

- Modernize to Pyton 3 (with AI help)
- Build Mingw using CMake
- MSVC build disabled, but can be added as an option later
- Various related changes and fixes

peterNordin committed:
Try fix cache variable

peterNordin committed:
Try fix cache variable

peterNordin committed:
Fix development checkout, depth is problematic

For official builds always clean clone
Depth is problematic when switching branches later

peterNordin committed:
Fix development checkout, depth is problematic

For official builds always clean clone
Depth is problematic when switching branches later

peterNordin committed:
Additional improvements to clone step

Avoid detached head for tags

peterNordin committed:
Additional improvements to clone step

Avoid detached head for tags

peterNordin committed:
Optionally link qtwebkit and install files

To enable CI build which may not have these files

peterNordin committed:
Optionally link qtwebkit and install files

To enable CI build which may not have these files

robbr48 reviewed (commented):
It works on Ubuntu. On Windows it compiles, but when I try to run from within Qt Creator it loops for the executables in the wrong folder. I was able to run Hopsan outside of Qt, but for some reason all default components failed to load.

peterNordin committed:
Update ssp4c build for msvc and switch to release for mingw

Also handle add workaround for the new cmake requirement

peterNordin committed:
Fix command checks

peterNordin committed:
Use windows 2022 image to keep using VS 2022

peterNordin committed:
Update ssp4c with various fixes

peterNordin committed:
Force lib dir name for libs in ssp4c

peterNordin committed:
Use cmake configuration package for ssp4c

peterNordin committed:
Fix and cleanup ssp4c related forward declarations

On MSVC struct and class can not be mixed

peterNordin committed:
Add ssp4c and tar.xz to gitignore

peterNordin committed:
Use explicit target name, it makes cmake easier to read

peterNordin committed:
Enable multi-core build on MSVC for most targets

peterNordin committed:
Add ssp4c build to all deb packages
